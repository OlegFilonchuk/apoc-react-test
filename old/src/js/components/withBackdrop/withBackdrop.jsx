import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Backdrop from '../Backdrop/Backdrop';

const getBackdrop = backdropProps => <Backdrop {...backdropProps} isVisible />;

export default (ComposedComponent, observedPropertyName) => {
  class DecoratedComponent extends Component {
    constructor(props) {
      super(props);

      /**
       * Hook into `onClose` event to detect internal state changes
       * in composed component
       */
      this.hasOnClose = typeof ComposedComponent.propTypes.onClose === 'function';

      if (this.hasOnClose) {
        this.onClose = (...args) => {
          this.setState(() => ({ isVisible: false }));

          /* eslint react/prop-types: 0 */
          this.props.onClose(...args);
        };
      }

      this.getHoc = this.getHoc.bind(this);
    }

    componentWillMount() {
      this.propsToState();
    }

    componentWillReceiveProps() {
      this.propsToState();
    }

    getHoc() {
      const { backdrop, ...props } = this.props;

      const backdropComponent = getBackdrop(backdrop);

      if (this.hasOnClose) {
        props.onClose = this.onClose;
      }

      const component = <ComposedComponent {...props} />;

      return (
        <div>
          {backdropComponent}
          {component}
        </div>
      );
    }

    propsToState() {
      this.setState((previousState, props) => ({ isVisible: props[observedPropertyName] }));
    }

    render() {
      return this.state.isVisible ? this.getHoc() : null;
    }
  }

  DecoratedComponent.propTypes = {
    backdrop: PropTypes.shape({
      ...Backdrop.propTypes
    })
  };

  DecoratedComponent.defaultProps = {
    backdrop: { ...Backdrop.defaultProps }
  };

  DecoratedComponent.displayName = `${ComposedComponent.displayName}WithBackdrop`;

  return DecoratedComponent;
};
