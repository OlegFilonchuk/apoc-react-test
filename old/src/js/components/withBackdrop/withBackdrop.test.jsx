import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';

import withBackdrop from './withBackdrop';

jest.useFakeTimers();

describe('withBackdrop HoC', () => {
  describe('general functionality', () => {
    const div = () => <div />;

    div.propTypes = {};

    it('should not render a component if no observed property provided', () => {
      const DivWithBackdrop = withBackdrop(div);
      const component = mount(<DivWithBackdrop />);

      expect(component.html()).toEqual(null);
    });

    it('should render <div /> with backdrop', () => {
      const DivWithBackdrop = withBackdrop(div, 'isOpen');
      const component = mount(<DivWithBackdrop isOpen />);

      expect(component).toMatchSnapshot();
    });
  });

  describe('onClose functionality', () => {
    const div = props => {
      setTimeout(() => props.onClose(), 0);

      return <div />;
    };

    div.propTypes = {
      onClose: PropTypes.func.isRequired
    };

    const onClose = jest.fn();

    const DivWithBackdrop = withBackdrop(div, 'isOpen');
    const component = mount(<DivWithBackdrop isOpen onClose={onClose} />);

    it('should not call onClose callback', () => {
      expect(onClose).not.toHaveBeenCalled();
    });

    describe('on close', () => {
      beforeAll(() => jest.runAllTimers());

      it('should hide element', () => {
        expect(component.html()).toEqual(null);
      });

      it('should call onClose callback', () => {
        expect(onClose).toHaveBeenCalled();
      });
    });
  });
});
