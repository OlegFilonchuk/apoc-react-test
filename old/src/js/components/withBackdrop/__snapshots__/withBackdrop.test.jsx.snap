// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withBackdrop HoC general functionality should render <div /> with backdrop 1`] = `
<undefinedWithBackdrop
  backdrop={
    Object {
      "className": "",
      "isVisible": false,
      "onClick": [Function],
      "zIndex": 15500,
    }
  }
  isOpen={true}
>
  <div>
    <Backdrop
      className=""
      isVisible={true}
      onClick={[Function]}
      zIndex={15500}
    >
      <div
        className="backdrop backdrop--visible"
        onClick={[Function]}
        style={
          Object {
            "zIndex": 15500,
          }
        }
      />
    </Backdrop>
    <div
      isOpen={true}
    >
      <div />
    </div>
  </div>
</undefinedWithBackdrop>
`;
